Helper/Api/Products.php

```diff
@@ -1087,65 +1092,58 @@ private function processProduct(
      * @param array $flat_product
      * @param Product $target
      * @param bool $shouldUpdate
-     * @return bool
+     * @return Product
      */
-    private function processImages(array $flat_product, Product $target, bool $shouldUpdate): bool
+    private function processImages(array $flat_product, Product $target, bool $shouldUpdate): Product
     {
-        if (isset($flat_product['main_image'])) {
-            $existingImage = null;
-
-            if ($target->getImage()) {
-                $existingImage = $this->getImageName($target->getImage());
-            }
-
-            if ($existingImage) {
-                $newImageName = $this->getImageName($flat_product['main_image']['big_url']);
-
-                if (
-                    $existingImage == 'no_selection'
-                    || !preg_match("/^{$newImageName}\_[0-9]+/", $existingImage)
-                ) {
-                    $shouldUpdate = $this->setGalleryImage(
-                        $flat_product['main_image']['big_url'],
-                        $target,
-                        true
-                    );
+        $galleryImages = $target->getMediaGalleryEntries();
+        $existingImagesArray = [];
+        foreach ($galleryImages as $entryId => $image) {
+            $mediaGalleryImage = $target->getMediaGalleryImages()->getItemById($image->getId());
+            if ($mediaGalleryImage) {
+                $skImageId = $mediaGalleryImage->getStorekeeperImageId();
+                $skImageIds = array_column($flat_product['product_images'], 'id');
+                /**
+                 * Look for storekeeper image id of current gallery image in array of images from SK backoffice
+                 * if current product does not match id or id is missing - remove gallery image
+                 */
+                if (!in_array($skImageId, $skImageIds)) {
+                    unset($galleryImages[$entryId]);
```

Comment:
generally removing items from a loop you are iterating over is not a good idea, better to copy over and do one update on product after the loop